pkg/runcfanotify/runcfanotify.go

```diff
@@ -0,0 +1,352 @@
+
+	for _, file := range runcPaths {
+		err = notify.Mark(unix.FAN_MARK_ADD, unix.FAN_OPEN_EXEC_PERM, unix.AT_FDCWD, file)
+		if err == nil {
+			log.Debugf("Checking %q: done", file)
+		} else {
+			log.Debugf("Checking %q: %s", file, err)
+		}
+	}
+
+	go n.run()
+
+	return n, nil
+}
+
+func monitorRuncInstance(bundleDir string, pidFile string, callback RuncNotifyFunc) error {
+	fanotifyFlags := uint(unix.FAN_CLOEXEC | unix.FAN_CLASS_CONTENT | unix.FAN_UNLIMITED_QUEUE | unix.FAN_UNLIMITED_MARKS)
+	openFlags := os.O_RDONLY | unix.O_LARGEFILE | unix.O_CLOEXEC
+
+	notify, err := fanotify.Initialize(fanotifyFlags, openFlags)
+	if err != nil {
+		return err
+	}
+
+	pidFileDir := filepath.Dir(pidFile)
+	log.Infof("runcfanotify: mark %s\n", pidFileDir)
+	err = notify.Mark(unix.FAN_MARK_ADD, unix.FAN_ACCESS_PERM|unix.FAN_EVENT_ON_CHILD, unix.AT_FDCWD, pidFileDir)
+	if err != nil {
+		log.Infof("Cannot mark %s: %s", bundleDir, err)
+	}
+	configJsonPath := filepath.Join(bundleDir, "config.json")
+	log.Infof("runcfanotify: mark ignore %s", configJsonPath)
+	err = notify.Mark(unix.FAN_MARK_ADD|unix.FAN_MARK_IGNORED_MASK, unix.FAN_ACCESS_PERM, unix.AT_FDCWD, configJsonPath)
+	if err != nil {
+		log.Infof("Cannot ignore %s/config.json: %s\n", bundleDir, err)
+	}
+
+	containerID := filepath.Base(filepath.Clean(bundleDir))
+
+	// f returns containerID, containerPID, error
+	f := func(notify *fanotify.NotifyFD, callback RuncNotifyFunc) (bool, error) {
+		data, err := notify.GetEvent()
+		if err != nil {
+			return false, fmt.Errorf("%w", err)
+		}
+
+		if data == nil {
+			return false, nil
+		}
+
+		defer data.Close()
+		if !data.MatchMask(unix.FAN_ACCESS_PERM) {
+			return false, fmt.Errorf("fanotify: unknown event")
+		}
+		defer notify.ResponseAllow(data)
+
+		pid := data.GetPID()
+
+		// Cannot use GetEvent(skipPIDs) because we need to call ResponseAllow
+		if pid == os.Getpid() {
+			return false, nil
+		}
+
+		path, err := data.GetPath()
+		if err != nil {
+			return false, err
+		}
+		if path != pidFile {
+			return false, nil
+		}
+
+		commB, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/comm", pid))
+		if err != nil {
+			return false, err
+		}
+		comm := strings.TrimSuffix(string(commB), "\n")
+
+		if comm != "containerd-shim" {
+			return false, nil
+		}
+
+		// Unfortunately, Linux 5.4 doesn't respect ignore masks
+		// See fix in Linux 5.9:
+		// https://github.com/torvalds/linux/commit/497b0c5a7c0688c1b100a9c2e267337f677c198e
+		// Workaround: remove parent mask. We don't need it anymore :)
+		err = notify.Mark(unix.FAN_MARK_REMOVE, unix.FAN_ACCESS_PERM|unix.FAN_EVENT_ON_CHILD, unix.AT_FDCWD, pidFileDir)
+		if err != nil {
+			return false, nil
+		}
+
+		pidFileContent, err := ioutil.ReadAll(data.File())
+		if err != nil {
+			return false, err
+		}
+		containerPID, err := strconv.Atoi(string(pidFileContent))
+		if err != nil {
+			return false, err
+		}
+
+		pidfd, _, errno := unix.Syscall(unix.SYS_PIDFD_OPEN, uintptr(containerPID), 0, 0)
```

Comment:
It's only used to monitor container termination, right? 

If yes, could we use the fanotify to watch for the deletion of the pid or config.json file?

Reply:
We could: support for watching deletion with fanotify are added in Linux 5.1. Or we could use inotify to support older kernels.

That would not be synchronous because `FAN_DELETE_SELF` is not synchronous but `pidfd_open()`'s notification is not synchronous either.

But is the pidfile immediately deleted by containerd-shim? I feel that pidfd_open is safer and more independent from containerd/runc's implementation.

Reply:
My only concern about pidfd_open is that is requires a more recent kernel and fanotify. Is it possible to check for the config.json removal?